src/grep.ts

```diff
@@ -0,0 +1,88 @@
+import * as exec from "@actions/exec";
+
+export type GrepSyntax = "extended" | "fixed" | "basic" | "perl";
+
+function grepSyntaxOption(syntax: GrepSyntax): string {
+  switch (syntax) {
+    case "extended":
+      return "--extended-regexp";
+    case "fixed":
+      return "--fixed-strings";
+    case "basic":
+      return "--basic-regexp";
+    case "perl":
+      return "--perl-regexp";
+  }
+}
+
+export type GrepResult =
+  | {
+      tag: "match";
+      input: string;
+      path: string;
+      line: number;
+    }
+  | { tag: "nomatch"; input: string; message: string };
+
+export async function grep(
+  syntax: GrepSyntax,
+  pattern: string,
+  files: string[],
+  silent?: boolean
+): Promise<GrepResult[]> {
+  let stdout = "";
+
+  // Tricks grep into always adding <file>: by ensuring more than one file arg
+  files.push("/dev/null");
+
+  const grepArgs = [
+    "--line-number",
+    "--color=never",
+    grepSyntaxOption(syntax),
+    pattern,
+  ].concat(files);
+
+  await exec.exec("grep", grepArgs, {
+    listeners: {
+      stdout: (data: Buffer) => {
+        stdout += data.toString();
+      },
+    },
+    ignoreReturnCode: true,
+    silent: silent || false,
+  });
+
+  return parseGrep(stdout);
+}
+
+// Exported for testing
+export function parseGrep(stdout: string): GrepResult[] {
+  // TODO Naive newline split (no Windows)
+  return stdout.split("\n").map(parseGrepLine);
+}
+
+function parseGrepLine(input: string): GrepResult {
+  const regex = /^(?<path>[^:]+):(?<line>[0-9]+):.*$/;
+  const match = input.match(regex);
+
+  if (match === null) {
+    return {
+      tag: "nomatch",
+      input,
+      message: `"${input}" did not match`,
+    };
+  }
+
+  const path = match.groups.path;
+  const line = parseInt(match.groups.line, 10);
+
+  if (isNaN(line)) {
+    return {
+      tag: "nomatch",
+      input,
+      message: `Line (${match.groups.line}) did not parse as Integer`,
```

Comment:
If this is expected to happen it would be useful to have a test case, and if it's unexpected, is this instead an error condition?

Reply:
Yeah it seems like something to `throw` on; `"nomatch"` is different because the matcher worked and didn't find anything.

Reply:
> "nomatch" is different because the matcher worked and didn't find anything

Just to clarify, if the matcher doesn't find anything it won't report any results. It only reports a `{tag:nomatch}` result if `grep` found something, but the output didn't match the `file:line:content` regex. That is just as unexpected as a non-integer line component, except for the fact that something (`@github/exec`?) causes a single trailing empty line to come across to `parseGrepLine`. I chose to capture that as a `nomatch` that I later filter out. I guess the (better?) alternative would be to filter out the empty line on the way in. I'll look into that now.

Reply:
OK. The last commit should make this much better.